core/src/main/java/org/apache/spark/shuffle/api/TransferrableWritableByteChannel.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+import org.apache.spark.annotation.Private;
+
+/**
+ * :: Private ::
+ * Represents an output byte channel that can copy bytes from input file channels to some
+ * arbitrary storage system.
+ * <p>
+ * This API is provided for advanced users who can transfer bytes from a file channel to
+ * some output sink without copying data into memory. Most users should not need to use
+ * this functionality; this is primarily provided for the built-in shuffle storage backends
+ * that persist shuffle files on local disk.
+ * <p>
+ * For a simpler alternative, see {@link ShufflePartitionWriter}.
+ *
+ * @since 3.0.0
+ */
+@Private
+public interface TransferrableWritableByteChannel extends Closeable {
+
+  /**
+   * Copy all bytes from the source readable byte channel into this byte channel.
+   * <p>
+   * This method should block until all of the bytes from the source (that is, up until
+   * numBytesToTransfer) are available in the output storage layer.
+   *
+   * @param source File to transfer bytes from. Do not call anything on this channel other than
```

Comment:
We discussed why this can't return a `WritableByteChannel` in this thread: https://github.com/palantir/spark/pull/535#discussion_r280193123

Reply:
Let me try to summarize the discussion from that thread here so that we have some record of the decision in the mainline repository.

Basically this comes from a need to block the `close` method from closing the underlying channel, but still keeping the optimization of `FileChannel#transferFrom(otherFileChannel)`.

We can't return `WritableByteChannel` here, because `FileChannel#transferTo(writableByteChannel)` needs to specifically receive a `WritableByteChannel` that is an instance of `FileChannel`. But if the `WritableByteChannel` is indeed an instance of `FileChannel`, then `FileChannel#close()` will force the underlying channel to close, which is not what we want to do particularly for the local disk storage implementation given here.

I agree that the abstraction here is convoluted. I proposed having an abstraction on top of the input `FileChannel` to prevent the plugin system from calling arbitrary methods on the input, but we decided that wasn't necessary in https://github.com/palantir/spark/pull/535#discussion_r280282920.

Reply:
> Basically this comes from a need to block the close method from closing the underlying channel

I see. Still this is a really convoluted interface, which makes me more strongly prefer that this be hidden as an optimization in the internal, local file implementation, instead of being exposed in the public API.

If really exposing this in the public API, I think it would be better to have an interface that makes it clear what's the intent here: provide a `FileChannel` (and not any other kind of channel) that, depending on the plugin, should not be closed. So something like:

```
interface FileChannelWrapper {

  FileChannel openChannel();

  boolean shouldClose();

}
```

Plugins by default would just return a `null` wrapper, and then the code calling this could do:

```
FileChannel channel = transferToEnabled ? writer.fileChannelWrapper() : null;
if (channel != null) {
  Utils.copyBlahBlahBlah();
  if (channel.shouldClose()) channel.close();
} else {
  // non-nio code path.
}
```

It's not great, but it's (i) less code than you have here and (ii) much clearer what's the actual intent of this API, without the restrictions about what methods should and should not be called.

Reply:
Another option is to just return the `FileChannel` from `ShufflePartitionWriter`, and have the shuffle code never close the returned channel. Instead, the `ShuffleMapOutputWriter` would do it when it's safe to do so.

That's a little awkward though since it would basically codify into the interface what is just a suggestion in `openStream()` (where it says you should override `close()` if you don't want the underlying stream to be closed).

But that kinda gives me an idea for tweaking the above interface a bit; have this in `ShufflePartitionWriter` instead:

```
FileChannel openChannel();

default void closeChannel(FileChannel channel) = { channel.close(); }
```

If you do the same for the `openStream()` method (i.e. have a `closeStream()`), then the API is symmetrical, and you avoid the use of stream wrappers to block the call to `close()`.

Reply:
I went with a different sort of abstraction that captures the spirit of this idea in the latest patch. Please take a look and let me know what you think.

Reply:
Ah I didn't see your follow-up comment. I'm not as fond of passing the resource back to the plugin to ask the plugin to close it? I'm not sure, that isn't a very common programming idiom with respect to streams and resources - usually the `close` method is attached to the closeable resource itself.

I liked the channel wrapper abstraction. We used to have the `close` method attached to the `ShufflePartitionWriter` and only closed the partition writer itself and not the channel, but that seemed prone to error on the plugin writer's side - since the partition writer returns a closeable resource itself, should the plugin expect the closeable resource it returns to be closed, or should it expect `ShufflePartitionWriter#close` to close the resource? Making `openStream` and `openChannel` return closeable resources is clearer with respect to the lifecycle of the returned resource.

Reply:
Commenting here to keep context, even though this code doesn't exist anymore.

I see you implemented basically my first suggestion. I think that if you instead make `ShufflePartitionWriter` implement `Closeable`, you can achieve the same goal with less things for people to implement. (`openChannel()` would just return a `WritableByteChannel`, and `close()` would close whatever was opened by either `open` call.)

It could even simplify some code in `BypassMergeSortShuffleWriter` since now the `close()` call would be the same regardless of whether it's a stream or a channel.

But that's minor now, and I'm just trying to reduce the amount of code needed for this optimization to work. Current version seems fine otherwise.

Reply:
I prefer this approach over making `ShufflePartitionWriter` implement `Closeable`, because `ShufflePartitionWriter` would return a `Closeable` resource (`OutputStream` or `WritableByteChannel`) that itself is not closed by the caller. That's why we originally went with `TransferrableWritableByteChannel` which itself is `Closeable` and doesn't return any `Closeable` objects. I think this is an in-between version, but it's unfortunate that `WritableByteChannelWrapper` itself returns a `WritableByteChannel` that itself isn't closed.